File: backend/app/schemas/legal.py
"""Pydantic schemas for Legal Corpus and Statute Retrieval."""

from typing import Optional, List
from pydantic import BaseModel, Field


class LegalActOut(BaseModel):
    id: int
    name: str
    short_name: str
    year: int
    jurisdiction: str
    domain: str
    source: Optional[str] = None
    version: str
    created_at: str


class LegalSectionOut(BaseModel):
    id: int
    act_id: int
    act_short_name: Optional[str] = None
    section_number: str
    title: Optional[str] = None
    text: str
    domain: str
    language: str
    source_reference: Optional[str] = None
    created_at: str


class RetrievalMatch(BaseModel):
    act: str
    section: str
    title: Optional[str] = None
    relevant_text: str
    plain_language_summary: Optional[str] = None
    confidence: float
    source_reference: Optional[str] = None
    source_name: Optional[str] = None
    source_url: Optional[str] = None
    official_source_url: Optional[str] = None
    source_authority: Optional[str] = None
    source_type: Optional[str] = None
    historical_reference: Optional[str] = None
    state: Optional[str] = "All"
    domain: Optional[str] = "general"
    status: Optional[str] = "CURRENT"
    last_verified: Optional[str] = None
    why_applies: Optional[str] = None


class RetrievalResponseData(BaseModel):
    status: str  # "success" or "insufficient_confidence"
    matches: List[RetrievalMatch]
    state_verified: bool = True
    state_note: Optional[str] = None


class IntakeRequest(BaseModel):
    text: str = Field(min_length=10, max_length=10000)
    language: Optional[str] = None


class CaseAnalysisResponse(BaseModel):
    case_id: int
    language: str
    domain: str
    subdomain: Optional[str] = None
    state: Optional[str] = None
    city: Optional[str] = None
    incident_date: Optional[str] = None

    confidence: str

    facts: dict
    missing_information: list[str]

    legal_sources: list[dict]

    rights: list[dict]

    evidence: list[dict]

    action_steps: list[dict]

    blocked_claims: list[str]

File: frontend/src/types/legal.ts
export type LawStatus = "CURRENT" | "ACTIVE" | "HISTORICAL" | "REPEALED" | (string & {});

/** A single statutory match returned by retrieval / search (mirrors RetrievalMatch). */
export interface RetrievalMatch {
  act: string;
  section: string;
  title?: string | null;
  relevant_text: string;
  plain_language_summary?: string | null;
  confidence: number;
  source_reference?: string | null;
  source_name?: string | null;
  source_url?: string | null;
  official_source_url?: string | null;
  source_authority?: string | null;
  source_type?: string | null;
  historical_reference?: string | null;
  state?: string | null;
  domain?: string | null;
  status?: LawStatus | null;
  last_verified?: string | null;
  verification_status?: string | null;
  why_applies?: string | null;
}

export interface RetrievalResponse {
  status: string;
  matches: RetrievalMatch[];
  state_verified: boolean;
  state_note?: string | null;
}

/* ---------------- Corpus (Legal Resources) ---------------- */

export interface CorpusStats {
  total_acts: number;
  total_sections: number;
  domains: Record<string, number>;
}

export interface CorpusAct {
  id: number;
  name: string;
  short_name: string;
  year: number;
  jurisdiction: string;
  domain: string;
  description?: string | null;
  source_url?: string | null;
  section_count: number;
}

export interface CorpusSection {
  id: number;
  act_id: number;
  act_short_name?: string;
  section_number: string;
  title?: string | null;
  text: string;
  domain: string;
  language: string;
}

export interface CorpusSearchResult {
  query: string;
  domain?: string | null;
  status: string;
  total: number;
  results: RetrievalMatch[];
}

/** Directory rows are dynamic DB rows; render defensively. */
export interface CorpusAuthority {
  id?: number;
  name?: string;
  domain?: string;
  jurisdiction?: string;
  website?: string;
  official_url?: string;
  helpline?: string;
  description?: string;
  [key: string]: unknown;
}

export interface CorpusJudgment {
  id?: number;
  case_name?: string;
  title?: string;
  citation?: string;
  court?: string;
  year?: number | string;
  summary?: string;
  domain?: string;
  url?: string;
  official_url?: string;
  [key: string]: unknown;
}

export interface CorpusVerifyIssue {
  check: string;
  severity: string;
  detail: string;
}

export interface CorpusVerifyReport {
  passed: boolean;
  summary: string;
  issues: CorpusVerifyIssue[];
}
